syntax = "proto3";

import "nanopb.proto";

message TimeStamp
{
    int32 sec = 1;
    uint32 nanosec = 2;
}

message TwistCmd
{
    TimeStamp time = 1;
    float v = 2;
    float w = 3;
}

message LaserScan
{
    TimeStamp time = 1;
    float angle_min = 2;
    float angle_max = 3;
    float angle_increment = 4;
    float time_increment = 5;
    float scan_time = 6;
    float range_min = 7;
    float range_max = 8;
    repeated float ranges = 9 [ (nanopb).max_count = 12 ];
    repeated float intensities = 10 [ (nanopb).max_count = 12 ];
}

message JointStates
{
    TimeStamp time = 1;
    repeated string name = 2
      [ (nanopb).max_count = 2, (nanopb).max_length = 15 ];
    repeated double position = 3 [ (nanopb).max_count = 2 ];
    repeated double velocity = 4 [ (nanopb).max_count = 2 ];
    repeated double effort = 5 [ (nanopb).max_count = 2 ];
}

message IMU
{
    TimeStamp time = 1;
    float gyro_x = 2;
    float gyro_y = 3;
    float gyro_z = 4;
    float accel_x = 5;
    float accel_y = 6;
    float accel_z = 7;
}

message UdpPacket
{
    repeated LaserScan laser = 1 [ (nanopb).max_count = 10 ];
    optional JointStates joint_states = 2;
    optional TwistCmd cmd_vel = 3;
    optional IMU imu = 4;
}
